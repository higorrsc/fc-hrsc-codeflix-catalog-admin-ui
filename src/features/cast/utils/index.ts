import { CastMemberType } from '../types';

export function getCastMemberTypeOptions() {
  return Object.values(CastMemberType)
    .filter((value) => typeof value === 'number')
    .map((value) => ({
      value,
      label:
        CastMemberType[value as CastMemberType].charAt(0) +
        CastMemberType[value as CastMemberType].slice(1).toLowerCase(),
    }));
}
