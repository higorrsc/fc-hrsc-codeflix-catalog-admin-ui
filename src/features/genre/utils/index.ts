import { Category } from '../../category/types';
import { Genre } from '../types';

export const mapGenreToForm = (genre: Genre) => {
  return {
    id: genre.id,
    name: genre.name,
    is_active: genre?.is_active,
    categories_id: genre.categories?.map((category: Category) => category.id),
  };
};
