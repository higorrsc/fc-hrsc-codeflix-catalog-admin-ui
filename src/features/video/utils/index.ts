import { Video, VideoPayload } from 'src/features/video/types';

export const mapVideoToForm = (video: Video): VideoPayload => {
  return {
    id: video.id,
    title: video.title,
    description: video.description,
    year_launched: video.year_launched,
    opened: video.opened,
    rating: video.rating,
    duration: video.duration,
    genres_id: video.genres?.map((genre) => genre.id),
    categories_id: video.categories?.map((category) => category.id),
    cast_members_id: video.cast_members?.map((cast_member) => cast_member.id),
  };
};
