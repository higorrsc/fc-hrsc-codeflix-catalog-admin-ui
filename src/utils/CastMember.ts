import { CastMemberType } from '../types/CastMember';

const castMemberTypeLabelMap: Record<CastMemberType, string> = {
  [CastMemberType.ACTOR]: 'Actor',
  [CastMemberType.DIRECTOR]: 'Director',
};

export function getCastMemberTypeLabel(type: CastMemberType): string {
  return castMemberTypeLabelMap[type] || 'Unknown';
}

export const castMemberTypeOptions = [
  { value: CastMemberType.ACTOR, label: 'Actor' },
  { value: CastMemberType.DIRECTOR, label: 'Director' },
];
