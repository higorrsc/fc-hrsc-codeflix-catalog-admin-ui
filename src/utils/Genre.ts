import { Category } from '../types/Category';
import { Genre } from '../types/Genre';

export const mapGenreToForm = (genre: Genre) => {
  return {
    id: genre.id,
    name: genre.name,
    is_active: genre?.is_active,
    categories_id: genre.categories?.map((category: Category) => category.id),
  };
};
