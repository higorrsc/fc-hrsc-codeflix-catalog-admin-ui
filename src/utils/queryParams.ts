import { ListOptions } from '../types/_shared'; // Import ListOptions if needed for typing map

/**
 * Maps specific frontend parameter names (keys often matching ListOptions)
 * to their corresponding backend query parameter names expected by the API.
 * We use Partial<Record<keyof ListOptions | string, string>> to allow mapping
 * ListOptions keys and potentially other string keys if needed later.
 */
const paramKeyMap: Partial<Record<keyof ListOptions | string, string>> = {
  perPage: 'per_page',
  // Add other mappings if needed
};

/**
 * Parses an object containing potential query parameters into a URL query string
 * suitable for API requests.
 *
 * - Accepts an object like `ListOptions` or any object with potential query params.
 * - Filters out parameters with `null` or `undefined` values.
 * - Explicitly ignores UI-specific keys like `rowsPerPage`.
 * - Maps keys according to `paramKeyMap` (e.g., 'perPage' becomes 'per_page').
 * - Converts remaining parameter values to strings.
 *
 * @template T - The type of the params object (e.g., ListOptions).
 * @param params - An object containing potential query parameters.
 * @returns A URL query string (without the leading '?'). Returns an empty string
 *          if no valid parameters are found after filtering and mapping.
 */
export function parseQueryParams<T extends Record<string, any>>(
  params: T
): string {
  const queryParams = new URLSearchParams();

  Object.entries(params).forEach(([key, value]) => {
    // 1. Skip null/undefined/empty string for search
    if (
      value === null ||
      value === undefined ||
      (key === 'search' && value === '') // Also skip empty search strings
    ) {
      return;
    }

    // 2. Skip known UI-specific keys
    if (key === 'rowsPerPage') {
      return;
    }

    // 3. Determine the correct key name for the API
    const mappedKey = paramKeyMap[key as keyof typeof paramKeyMap] || key;

    // 4. Convert the value to a string and append
    queryParams.append(mappedKey, String(value));
  });

  return queryParams.toString();
}
