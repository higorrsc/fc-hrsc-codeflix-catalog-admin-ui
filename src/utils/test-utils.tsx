import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

import { ThemeProvider } from '@mui/system';
import { configureStore } from '@reduxjs/toolkit';
import { SnackbarProvider } from 'notistack';
import { BrowserRouter } from 'react-router-dom';
import type { RootState } from '../app/store';
import { appTheme } from '../config/theme';
import { apiSlice } from '../features/api/apiSlice';
import { castMembersApiSlice } from '../features/cast/castMemberSlice';
import { categoriesApiSlice } from '../features/categories/categorySlice';
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: ReturnType<typeof configureStore>;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        [categoriesApiSlice.reducerPath]: apiSlice.reducer,
        // @ts-ignore
        [castMembersApiSlice.reducerPath]: apiSlice.reducer,
      },
      // Adding the api middleware enables caching, invalidation, polling,
      // and other useful features of `rtk-query`.
      middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(apiSlice.middleware),
    }),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <React.StrictMode>
      <BrowserRouter>
        <Provider store={store}>
          <ThemeProvider theme={appTheme}>
            <SnackbarProvider maxSnack={3}>{children}</SnackbarProvider>
          </ThemeProvider>
        </Provider>
      </BrowserRouter>
    </React.StrictMode>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
