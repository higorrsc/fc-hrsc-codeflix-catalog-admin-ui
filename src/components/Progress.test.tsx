import { render, screen } from '../utils/test-utils';
import { LinearProgressWithLabel } from './Progress';

describe('Test LinearProgressWithLabel Component', () => {
  it('should render correctly with default value (0%)', () => {
    const { asFragment } = render(<LinearProgressWithLabel value={0} />);

    // Check for the progress bar role
    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toBeInTheDocument();

    // Check for the label
    expect(screen.getByText('0%')).toBeInTheDocument();

    // Snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with a specific value', () => {
    render(<LinearProgressWithLabel value={50} />);

    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toBeInTheDocument();
    expect(progressBar).toHaveAttribute('aria-valuenow', '50');

    expect(screen.getByText('50%')).toBeInTheDocument();
  });

  it('should render correctly with value 0', () => {
    render(<LinearProgressWithLabel value={0} />);

    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toBeInTheDocument();
    expect(progressBar).toHaveAttribute('aria-valuenow', '0');

    expect(screen.getByText('0%')).toBeInTheDocument();
  });

  it('should render correctly with value 100', () => {
    render(<LinearProgressWithLabel value={100} />);

    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toBeInTheDocument();
    expect(progressBar).toHaveAttribute('aria-valuenow', '100');

    expect(screen.getByText('100%')).toBeInTheDocument();
  });

  it('should round the value for the label display', () => {
    render(<LinearProgressWithLabel value={75.7} />);
    expect(screen.getByText('76%')).toBeInTheDocument(); // 75.7 rounds to 76
  });
});
