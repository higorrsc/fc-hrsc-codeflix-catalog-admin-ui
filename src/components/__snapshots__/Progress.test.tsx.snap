// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LinearProgressWithLabel Component should render correctly with default value (0%) 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-70qvj9"
  >
    <div
      class="MuiBox-root css-1bhivts"
    >
      <span
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-1txyin7-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-l16vtb-MuiLinearProgress-bar1"
          style="transform: translateX(-100%);"
        />
      </span>
    </div>
    <div
      class="MuiBox-root css-pazuet"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
      >
        0%
      </p>
    </div>
  </div>
</DocumentFragment>
`;
